lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/projectcalico/libcalico-go/lib/net"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	BlockAffinityResourceName = "BlockAffinities"
+	BlockAffinityCRDName      = "blockaffinities.crd.projectcalico.org"
+)
+
+func NewAffinityBlockClient(c *kubernetes.Clientset, r *rest.RESTClient) K8sResourceClient {
+	// Create a resource client which manages k8s CRDs.
+	rc := customK8sResourceClient{
+		clientSet:       c,
+		restClient:      r,
+		name:            BlockAffinityCRDName,
+		resource:        BlockAffinityResourceName,
+		description:     "Calico IPAM block affinities",
+		k8sResourceType: reflect.TypeOf(apiv3.BlockAffinity{}),
+		k8sResourceTypeMeta: metav1.TypeMeta{
+			Kind:       apiv3.KindBlockAffinity,
+			APIVersion: apiv3.GroupVersionCurrent,
+		},
+		k8sListType:  reflect.TypeOf(apiv3.BlockAffinityList{}),
+		resourceKind: apiv3.KindBlockAffinity,
+	}
+
+	return &affinityBlockClient{rc: rc}
+}
+
+// Implements the api.Client interface for AffinityBlocks.
+type affinityBlockClient struct {
+	rc customK8sResourceClient
+}
+
+func toV1(kvpv3 *model.KVPair) *model.KVPair {
+	cidrStr := kvpv3.Value.(*apiv3.BlockAffinity).Annotations["projectcalico.org/cidr"]
+	host := kvpv3.Value.(*apiv3.BlockAffinity).Annotations["projectcalico.org/host"]
+	_, cidr, err := net.ParseCIDR(cidrStr)
+	if err != nil {
+		panic(err)
+	}
+	state := model.BlockAffinityState(kvpv3.Value.(*apiv3.BlockAffinity).Spec.State)
+	return &model.KVPair{
+		Key: model.BlockAffinityKey{
+			Host: host,
+			CIDR: *cidr,
+			UID:  &kvpv3.Value.(*apiv3.BlockAffinity).UID,
+		},
+		Value: &model.BlockAffinity{
+			State: state,
+		},
+		Revision: kvpv3.Revision,
+	}
+}
+
+func v3Fields(k model.Key) (name, cidr, host string) {
+	host = k.(model.BlockAffinityKey).Host
+	cidr = fmt.Sprintf("%s", k.(model.BlockAffinityKey).CIDR)
+	h := sha1.New()
+	h.Write([]byte(fmt.Sprintf("%s+%s", host, cidr)))
+	name = fmt.Sprintf("%s-%sc", host, hex.EncodeToString(h.Sum(nil))[:11])
+	return
+}
+
+func toV3(kvpv1 *model.KVPair) *model.KVPair {
+	name, cidr, host := v3Fields(kvpv1.Key)
+	state := kvpv1.Value.(*model.BlockAffinity).State
+	return &model.KVPair{
+		Key: model.ResourceKey{
+			Name: name,
+			Kind: apiv3.KindBlockAffinity,
+		},
+		Value: &apiv3.BlockAffinity{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       apiv3.KindBlockAffinity,
+				APIVersion: "crd.projectcalico.org/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            name,
+				ResourceVersion: kvpv1.Revision,
+				Annotations: map[string]string{
+					"projectcalico.org/host": host,
+					"projectcalico.org/cidr": cidr,
+				},
+			},
+			Spec: apiv3.BlockAffinitySpec{
+				State: string(state),
+			},
+		},
+		Revision: kvpv1.Revision,
+	}
+}
+
+func (c *affinityBlockClient) Create(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
```

Comment:
This `affinityBlockClieny` exists solely to do v1<->v3 and conversion, since the CRDs are in "v3" format and the IPAM code uses models.

We should clean this up, either by:
- Updating the IPAM code to use v3 style resources, and adjusting etcdv3 to to the conversion.
- Enhancing `customK8sResourceClient` to do the conversion for us, removing the need for so much boilerplate. 
- ???

Reply:
For now, I suggest we allow a bit of duplication to reduce the need for additional interfaces in the backend.... I think the code will be more readable this way.